### ISSUE TITLE
& might be a bad choice for de-interpolation

### ISSUE THREAD

**Issue by Roger-luo** [thautwarm/MLStyle.jl#160]
it has different pred rule as `$` so it is not very intuitive to use it sometimes

```julia
julia> ex = :(&A.B(a, b, c))
:(&(A.B(a, b, c)))

julia> ex.head
:&

julia> ex.args
1-element Vector{Any}:
 :(A.B(a, b, c))

julia> ex = :(:($A.B(a, b, c)))
:($(Expr(:quote, :(($(Expr(:$, :A))).B(a, b, c)))))
```

this as a result caused us eval the entire call instead of the object in the following case

```julia
function Tree.print_node(io::IO, node::SpatialModulationType)
    @match node begin
        &SpatialModulationType.Global => print(io, "Global")
        &SpatialModulationType.ScaledLocations(_) =>  print(io, "ScaledLocations")
    end
end
```

`&` is still a nice and well-adopted symbol tho, though maybe we just need to find a way to fix the pred by matching more Julia expression patterns

**Comment by thautwarm** [comment]
Yes, we need to consider this in our new rewrite.

We can learn from Python's design: https://peps.python.org/pep-0636/

`A.b` == matching by value is pretty okay as the introduction in Python doesn't cause issues in practice. We don't need pin operators in this case. 

However, I'm not sure how to support matching a value that comes from the local scope, or we just do not support this? Actually, this is only supported in pattern matching languages that support pin operators. See this:

```julia
function func(local_var)
    @match value begin
        &local_var => ...
    end
end
```

**Comment by Roger-luo** [comment]
True, I think we still need a distinguish of local/global scope just because this is not available inside a Julia macro, but if somehow we can get macro caller scope, this would not be necessary the rule is the same as other cases if the name is defined then match the value, instead of assigning a new variable. 

It seems we have no choice but to have something like this because the support has to be implemented as a macro instead of after-scope gets resolved. It's just we need to think about the operators pred more so perhaps need something else other than `&`

**Comment by Roger-luo** [comment]
I think we should just use `$`, the meaning of this is actually clear outside quotes, e.g

```julia
function Tree.print_node(io::IO, node::SpatialModulationType)
    @match node begin
        $SpatialModulationType.Global => print(io, "Global")
        $SpatialModulationType.ScaledLocations(_) =>  print(io, "ScaledLocations")
    end
end
```

just means we insert a value from outside to the pattern expression, so it will reference to whatever `SpatialModulationType` points to within this scope. Within a quote expression, we just need to do it twice

```julia
@match ex begin
    :($($x) + 1) => true
    _ => false
end
```

this means we reference an external value and then insert it into the pattern, if `x = 1` this is equivalent to writing

```julia
@match ex begin
    :($(1) + 1) => true
    _ => false
end
```

**Comment by thautwarm** [comment]
I think you are correct. I'm checking expressions created from Julia parsers could support this.

**Comment by jariji** [comment]
Here is a use case for `&`. Suppose I want to rewrite `x + x` to `2x` for any `x`.

```jl
julia> @match :(x + x) :($e + $f) && if e == f end => :(2($e))
:(2x)

julia> @match :(x + x) :($e + $(&e)) => :(2($e))
:(2x)
```

I don't think either of these syntaxes is especially pretty, but the `&` is probably nicer.
